server/elasticsearch/src/main/java/org/hawkular/apm/server/elasticsearch/TraceServiceElasticsearch.java

```diff
@@ -132,14 +134,69 @@ public Trace getFragment(String tenantId, String id) {
         return ret;
     }
 
+    protected List<Trace> getFragmentsForTraceId(String tenantId, String traceId) {
+        List<Trace> ret = new ArrayList<Trace>();
+
+        String index = client.getIndex(tenantId);
+
+        try {
+            RefreshRequestBuilder refreshRequestBuilder = client.getClient().admin().indices().prepareRefresh(index);
+            client.getClient().admin().indices().refresh(refreshRequestBuilder.request()).actionGet();
+
+            BoolQueryBuilder query = QueryBuilders.boolQuery()
+                    .must(QueryBuilders.termQuery("traceId", traceId));
+
+            SearchRequestBuilder request = client.getClient().prepareSearch(index)
+                    .setTypes(TRACE_TYPE)
+                    .setSearchType(SearchType.DFS_QUERY_THEN_FETCH)
+                    .setQuery(query)
+                    .addSort(ElasticsearchUtil.TIMESTAMP_FIELD, SortOrder.ASC);
+
+            SearchResponse response = request.execute().actionGet();
+            if (response.isTimedOut()) {
+                msgLog.warnQueryTimedOut();
+            }
+
+            for (SearchHit searchHitFields : response.getHits()) {
+                try {
+                    ret.add(mapper.readValue(searchHitFields.getSourceAsString(),
+                            Trace.class));
+                } catch (IOException e) {
+                    msgLog.errorFailedToParse(e);
+                }
+            }
+
+            if (msgLog.isTraceEnabled()) {
+                msgLog.tracef("Query fragments with traceId[%s] is: %s", traceId, ret);
+            }
+        } catch (org.elasticsearch.indices.IndexMissingException ime) {
+            // Ignore, as means that no traces have
+            // been stored yet
+            if (msgLog.isTraceEnabled()) {
+                msgLog.tracef("No index found, so unable to retrieve traces");
+            }
+        } catch (org.elasticsearch.action.search.SearchPhaseExecutionException spee) {
+            // Ignore, as occurs when mapping not established (i.e. empty
+            // repository) and performing query with a sort order
+            if (msgLog.isTraceEnabled()) {
+                msgLog.tracef("Failed to get fragments", spee);
+            }
+        }
+
+        return ret;
+    }
+
     @Override
     public Trace getTrace(String tenantId, String id) {
-        Trace ret = getFragment(tenantId, id);
+        Trace ret = null;
+        List<Trace> fragments = getFragmentsForTraceId(tenantId, id);
+
+        if (!fragments.isEmpty()) {
+            ret = fragments.stream().filter(f -> f.getTraceId().equals(id)).findFirst().get();
```

Comment:
Could you please move this after `List<Trace> fragments = getFragmentsForTraceId(tenantId, id);` with `ret` definition. And change the if to `ret != null` and remove `ret != null` from the for statement?